libs/@local/hql/README.md

```diff
@@ -0,0 +1,51 @@
+# HQL - HASH Query Language
+
+This is the compiler, planner and executor for the HASH Query Language (HQL). HQL is a functional programming language that is created using a subset of JSON.
+
+## Language
+
+Language design documentation and a specification of the language will be provided in the future.
+
+The CST of the language is very basic, only consisting of three different constructs:
+
+- Function Invocation: A function invocation has multiple possible forms:
+  - `[name, ...args]`
+  - `{"fn": name, "args": args}`
+- Constants: Constants are simple values that are not functions, they are defined using:
+  - `{"const": value}`, this will convert the JSON type to the corresponding `hql` type.
```

Comment:
In which scenario `type` is not known?

Reply:
it's in the case of parsing, e.g. you have a constant of `1`, if you say:

* `{"const": 1}`, it will be converted to the equivalent JSON representation (so just `Number`) or it being inferred.
* `{"const": 1, "type": "Int"}`, will take the value in `const`, and try to assign it to `Int`, which would (statically) fail for e.g. `1.1`.

This is mostly an affordance to the user. I reckon that in most cases you don't need to specify the type.

It's very similar to rust, meaning:

```rs
let x: _ = 2;
// vs.
let x: u8 = 2;
```

Here, if we have `_`, if there's no other usage it would default to `i32`, but if we have a function of e.g. `something(a: u8)`, it would automatically constrain `x` to be `u8` instead.

Reply:
In that case let's encode it as `{type: "constant", value: Value, type?: Type }` instead. The CST for the above, btw, does not omit `type`, it's basically `type: "inferred"` or something like that.

Reply:
Yes that's correct, that is why it is represented by an `Option` in the CST itself, I didn't want to make any type assumptions in the CST, if type checking is part of the AST instead.

I also wanted to keep the syntax terse, which is why I have been using `{"const"}` (or `constant`) instead of `{"type": "constant", "value": ...}`, it also opens up the word `type` to be used for the type instead.

Reply:
The `type: "constant"` is probably not needed, yes, but it helps deserializing, but it doens't really matter as we implement the deserializing anyway.